project-minimization-core/src/main/kotlin/org/plan/research/minimization/core/algorithm/graph/hierarchical/GraphHierarchyProducer.kt

```diff
@@ -0,0 +1,106 @@
+import org.plan.research.minimization.core.model.PropertyTester
+import org.plan.research.minimization.core.model.graph.GraphEdge
+import org.plan.research.minimization.core.model.graph.GraphWithAdjacencyList
+
+import arrow.core.getOrElse
+import arrow.core.raise.ensureNotNull
+import arrow.core.raise.option
+
+import kotlinx.collections.immutable.mutate
+import kotlinx.collections.immutable.persistentMapOf
+
+private typealias GraphHierarchyPropertyTester<V, E, G, C> = PropertyTester<GraphHierarchicalDDContext<V, E, G, C>, CondensedVertex<V>>
+private typealias DdGraphResult<V, E, G, C> = DDAlgorithmResult<GraphHierarchicalDDContext<V, E, G, C>, CondensedVertex<V>>
+
+internal class GraphHierarchyProducer<V, E, G, C>(val propertyTester: GraphHierarchyPropertyTester<V, E, G, C>) :
+    HierarchicalDDGenerator<GraphHierarchicalDDContext<V, E, G, C>, CondensedVertex<V>>
+where V : DDItem,
+E : GraphEdge<V>,
+G : GraphWithAdjacencyList<V, E>,
+C : GraphContext<V, E, G> {
+    override suspend fun generateFirstLevel(context: GraphHierarchicalDDContext<V, E, G, C>) = option {
+        val condensedVertexSet = StrongConnectivityCondensation.compressGraph(context.graph)
+        val compressedGraph = CondensedGraph.from(condensedVertexSet)
+        val sinks = compressedGraph
+            .sinks
+
+        ensure(sinks.isNotEmpty())
+        HDDLevel(
+            context = context.copy(
+                currentLevel = sinks,
+                inactiveElements = persistentMapOf(),
+                condensedGraph = compressedGraph,
+            ),
+            propertyTester = propertyTester,
+            items = sinks,
+        )
+    }
+
+    override suspend fun generateNextLevel(minimizationResult: DdGraphResult<V, E, G, C>) =
+        option {
+            val deletedPropagated = minimizationResult.propagateDeleted()
+            val (propagatedContext, nextInactiveElements) = deletedPropagated.propagateActive().bind()
+            propagatedContext
+                .produceNextLevel(nextInactiveElements)
+                .bind()
+        }
+
+    private fun DdGraphResult<V, E, G, C>.propagateDeleted(): DdGraphResult<V, E, G, C> {
+        val graph = requireNotNull(context.condensedGraph)
+
+        val currentLevel = requireNotNull(context.currentLevel)
+        require(currentLevel.all { it is CondensedVertex<*> })
+        val deletedItems = (currentLevel as List<CondensedVertex<V>>).toSet() - items.toSet()
+        val deletedDependenciesCollector = DeletedDependenciesCollector<V, E>(deletedItems)
+        val deletedWithDependencies = deletedDependenciesCollector.visitGraph(graph)
+        val newContext = context.copy(
+            inactiveElements = context.inactiveElements.mutate { map -> deletedWithDependencies.forEach(map::remove) },
+            condensedGraph = graph.withoutNodes(deletedWithDependencies),
+        )
+        return DdGraphResult<V, E, G, C>(
+            newContext,
+            items,
+        )
+    }
```

Comment:
I'm sure that this should happen inside property testing

Reply:
I suppose there is no possible way to do it without propagating inactive and deleted elements

Reply:
In some terms -- yes, but it shouldn't be here anyway because this propagation should happen on each testing, not only here

Reply:
Yes and no. This exact propagation is needed in terms of hierarchy: we still need to support it in the hierarchy producer to get a correct result, since HDD's scope is trees. 

However, I agree that moving that logic into property tester would make things easier

Reply:
That is why I suggest to don't use HDD (btw the interface is designed not only for trees)